### PR TITLE
feat(alerts): oracle price jump vs swap fee

### PR DESCRIPTION
**Pull request by chapati23**
## Summary

New risk-vector alert: when a `MedianUpdated` event moves the oracle price by more than the pool's combined swap fee, arbitrageurs round-trip through the pool and extract the excess as LP losses. We now detect those jumps and page on them.

- **Warning** — jump is above `lpFee + protocolFee` but still within 10% of it (e.g. 10 bps fee, 10.5 bps jump).
- **Critical** — jump is ≥10% above the fee (e.g. 10 bps fee, 11 bps+ jump).

Mutually exclusive: a single jump matches exactly one tier. Boundary written as `jump * 10 ⋈ fee * 11` so the 10% cutoff is IEEE-754 stable for fees that aren't multiples of 10.

## Deploy order (pre-merge checklist)

Following the discipline established by PR #219 (indexer-first, no bridge fallback):

- [ ] Indexer schema + handler changes deployed to the hosted Envio indexer (`pnpm deploy:indexer`)
- [ ] Confirmed the new `Pool.lastOracleJumpBps` / `lastOracleJumpAt` / `lastMedianPrice` / `lpFee` / `protocolFee` fields respond in Hasura
- [ ] This PR merged — `metrics-bridge` auto-deploy will hit a ready schema

Merging before the indexer is ready crashes `metrics-bridge` on the `BridgePools` query.

## Design

Follows the PR #212 (Rebalance Effectiveness) pattern end-to-end:

1. **Indexer** (`sortedOracles.ts` → `MedianUpdated`): compute `|newMedian − prevMedian| / prevMedian × 10_000` as a 4dp bps string, stored on the `Pool` entity. Uses a separate `lastMedianPrice` field because `OracleReported` also overwrites `oraclePrice` with individual reporter values — using `oraclePrice` as "previous median" would compare a median against a reporter value.
2. **metrics-bridge**: three new gauges — `mento_pool_swap_fee_bps`, `mento_pool_oracle_jump_bps`, `mento_pool_oracle_jump_at`. `-1` sentinel on either fee skips `swap_fee_bps` publication (so the alert doesn't compare against a stale threshold when RPC failed at pool creation); a real `0 + 0` fee publishes and remains alert-eligible.
3. **Grafana** (`terraform/alerts/rules-fpmms.tf`): new `Oracle Price Jump` rule group with two rules sharing a `(time() - oracle_jump_at) < 600` recency gate so a single large jump clears 10 min after the event instead of holding the alert open until the next median.

No FX-weekend suppression — a Monday-open gap IS the LP-leakage event we want to catch, consistent with the existing un-suppressed `Oracle Down` rule.

## Test plan

- [x] `pnpm --filter @mento-protocol/indexer-envio test:coverage` — 319 passing, 10 new covering first-report, up/down jumps, sub-bps precision, warning and critical alert boundaries
- [x] `pnpm --filter @mento-protocol/metrics-bridge test` — 49 passing, 7 new for the three gauges (sentinel skip on each fee, real-zero publish path, float parsing, timestamp)
- [x] `terraform validate` + `terraform fmt -check` in `terraform/alerts/` — clean
- [x] `trunk check` — no issues
- [ ] Post-merge smoke: lower the critical multiplier from `* 11 / 10` to `* 5 / 10` on a staging pool, `terraform apply`, confirm Slack message lands in `#alerts-critical`, revert

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Adds new indexer fields and metrics/alerts that directly drive paging; incorrect deployment order or edge-case math could break the `metrics-bridge` GraphQL query or misclassify alert severity.
> 
> **Overview**
> Adds oracle *median-to-median jump* tracking to the indexer by extending `Pool` with `lastMedianPrice`, `lastOracleJumpBps`, and `lastOracleJumpAt`, computing the jump on `SortedOracles.MedianUpdated` via a new `computeOracleJumpBps` helper (with tests).
> 
> Updates `metrics-bridge` to query the new pool fields plus `lpFee`/`protocolFee`, publish new gauges for `mento_pool_swap_fee_bps`, `mento_pool_oracle_jump_bps`, and `mento_pool_oracle_jump_at` (skipping fee `-1` sentinels), and adds coverage for fee/jump edge cases and boundary behavior.
> 
> Introduces a new Terraform Grafana rule group `Oracle Price Jump` with mutually exclusive warning/critical rules that compare jump vs. swap fee using an IEEE-stable boundary (`jump * 10` vs `fee * 11`) and a 10-minute recency gate based on `oracle_jump_at`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b27199453724d76c97fde165a802f47146480504. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by chapati23** [comment]
@claude review

**Comment by chapati23** [comment]
cursor review

File: terraform/alerts/rules-fpmms.tf [hunk 1]
```diff
@@ -839,3 +839,208 @@ resource "grafana_rule_group" "fpmms_rebalancer" {
...
+        expr = join(" and ", [
+          "mento_pool_oracle_jump_bps",
+          "(mento_pool_oracle_jump_bps > mento_pool_swap_fee_bps)",
+          # Strict `<` upper bound: at exactly swap_fee × 1.10 the critical
+          # rule takes over. Keeping this bound in the warning expression
+          # (instead of deferring to critical-wins-on-overlap) means both
+          # severities can safely route to different channels.
+          "(mento_pool_oracle_jump_bps < mento_pool_swap_fee_bps * 1.10)",
+          "((time() - mento_pool_oracle_jump_at) < 600)",
+          "(mento_pool_swap_fee_bps > 0)",
```
[the remark this answers is not in the document]
Comment (chapati23) [reply terraform/alerts/rules-fpmms.tf]: Fixed in d4961d3 — swapped `> 0` for `>= 0` in both alert rules. The bridge already filters the `-1` sentinel by skipping `.set()` entirely (series absent), so every published value is a real fee and a 0-bps configuration now stays eligible to alert. Added a bridge regression test pinning the zero-fee publish path.

File: terraform/alerts/rules-fpmms.tf [hunk 2]
```diff
@@ -839,3 +839,208 @@ resource "grafana_rule_group" "fpmms_rebalancer" {
...
+      datasource_uid = var.prometheus_datasource_uid
+      relative_time_range {
+        from = local.instant_query_range_seconds
+        to   = 0
+      }
+      model = jsonencode({
+        refId = "A"
+        expr = join(" and ", [
+          "mento_pool_oracle_jump_bps",
+          "(mento_pool_oracle_jump_bps >= mento_pool_swap_fee_bps * 1.10)",
```
[the remark this answers is not in the document]
Comment (chapati23) [reply terraform/alerts/rules-fpmms.tf]: Fixed in d4961d3 — rewrote both thresholds as `jump * 10 ⋈ fee * 11`. `fee * 11` is integer-exact so the 10% boundary no longer depends on `* 1.10` float residue. Verified: the problematic fee=3/jump=3.3 case now routes to critical (was warning). Added comments on both rule blocks explaining the reformulation.

File: metrics-bridge/src/graphql.ts [hunk 3]
```diff
@@ -25,6 +25,10 @@ const BRIDGE_POOLS_QUERY = gql`
       lastEffectivenessRatio
       rebalanceLivenessStatus
       hasHealthData
+      lpFee
+      protocolFee
+      lastOracleJumpBps
```
[the remark this answers is not in the document]
Comment (chapati23) [reply metrics-bridge/src/graphql.ts:30]: Won't fix: the repo's chosen deploy discipline is indexer-first-then-bridge, not dual-read fallback. PR #212 added a schema-lag fallback for `lastEffectivenessRatio`; PR #219 (merged 2 days ago) **explicitly removed** that fallback once the indexer caught up. Re-adding a guard here would revert that decision.

The deploy order for this PR: (1) merge schema + handler changes to the indexer repo and wait for Envio to expose the new fields in Hasura, (2) merge this monorepo PR so `metrics-bridge` auto-deploys against the ready schema. I'll add a pre-merge checkbox to the PR description.

**Comment by chapati23** [comment]
@claude: addressed in d4961d3.

**Fixed**
- Zero-fee pool exclusion (`> 0` → `>= 0` in both rules). You're right — the sentinel is already filtered by the bridge skipping `.set()`, so every published value is real.

**Won't fix (nits 1 & 2)**
- `Number(lastOracleJumpAt)` conversion: consistent with `Number(pool.oracleTimestamp)`, `Number(pool.oracleExpiry)`, `Number(pool.deviationBreachStartedAt)`, etc. in the same file. Adding a comment here would noise up a pattern that's self-evident from the existing gauges.
- BigInt floor division truncates `10.49999…` to `10.4999`: agreed, truncation is safe for alerting (slightly fewer false positives, as you noted). Not changing.

**Comment by chapati23** [comment]
@cursor: addressed in d4961d3.

**Fixed**
- Zero-fee pool exclusion (`swap_fee_bps > 0` → `>= 0`). Your reasoning about the bridge already distinguishing sentinel vs real-zero was correct.
- Added bridge regression test locking in the real-zero path (`lpFee = 0, protocolFee = 0` → `swap_fee_bps = 0` published).
- Also: rewrote both boundary expressions as `jump * 10 ⋈ fee * 11` instead of `* 1.10` to remove IEEE-754 residue at the 10% cutoff (flagged by Codex in parallel).

**Won't fix: schema rollout dependency**
The repo's chosen deploy discipline is indexer-first-then-bridge, not dual-read fallback. PR #212 added a fallback for `lastEffectivenessRatio`; PR #219 explicitly removed it 2 days ago once the indexer caught up. Re-adding a guard here would revert that decision.

Deploy order for this PR, pinned to the PR description: (1) ship the schema + handler changes to the indexer repo and wait for Envio to expose the new fields in Hasura, (2) merge this monorepo PR so `metrics-bridge` auto-deploys against the ready schema.

File: metrics-bridge/src/metrics.ts [hunk 4]
```diff
@@ -183,6 +201,16 @@ export function updateMetrics(pools: PoolRow[]): void {
...
     }
+    // Swap fee — skip the `-1` sentinel the indexer writes when the initial
+    // RPC fetch at pool creation failed (rpc.ts:fetchFees). Without this gate
+    // the `Oracle Jump` alert would see a "0 bps" threshold and fire on the
+    // first real oracle movement. `-1` on either side means we can't trust
+    // the sum.
+    if (pool.lpFee >= 0 && pool.protocolFee >= 0) {
+      gauges.swapFeeBps.set(labels, pool.lpFee + pool.protocolFee);
+    }
+    gauges.oracleJumpBps.set(labels, fp(pool.lastOracleJumpBps));
```
[the remark this answers is not in the document]
Comment (chapati23) [reply metrics-bridge/src/metrics.ts:212]: Won't fix: the concern doesn't materialize under IEEE-754 round-to-nearest.

`parseFloat("3.3000")` does give an IEEE approximation slightly below 3.3 (`3.2999999999999998...`), but `3.3 * 10` rounds back to **33.0 exactly** — 33.0 is representable and closer to the infinite-precision product (32.99999999999999822…) than the next representable below (33 − 2⁻⁴⁶). So `jump * 10 >= fee * 11` → `33 >= 33` → true → critical fires. Verified across V8, which shares IEEE semantics with Go/PromQL.

Locked this in with a parameterised boundary regression test (`metrics.test.ts`, 022fb91) covering your exact case (fee=3, jump=3.3), sub- and super-boundary values, and the user's stated 10-bps-fee examples — mutual exclusion holds at every tested boundary. If a future gauge-unit or parseFloat swap breaks this, the test trips immediately.

Emitting as a scaled integer would eliminate all theoretical float concerns but adds complexity for a non-existent bug. Keeping the simpler form.

**Comment by chapati23** [comment]
@codex: acknowledged. The one inline finding (scaled-integer jump emission) is addressed in its thread — IEEE round-to-nearest keeps `jump * 10 >= fee * 11` correct at every boundary I can construct, and 022fb91 pins that with a parameterised test.

**Comment by chapati23** [comment]
@claude: thanks for the follow-up 🙏 — LGTM noted. Will hold merging until the indexer schema is live on Envio.

File: metrics-bridge/src/graphql.ts [hunk 5]
```diff
@@ -25,6 +25,10 @@ const BRIDGE_POOLS_QUERY = gql`
...
+      lastOracleJumpAt
```
[the remark this answers is not in the document]
Comment (chapati23) [reply metrics-bridge/src/graphql.ts:31]: Won't fix — duplicate of Cursor's finding on the same line, same thread resolution.

The repo's chosen deploy discipline is indexer-first-then-bridge, not dual-read fallback. PR #212 added a schema-lag fallback for `lastEffectivenessRatio`; PR #219 (merged 2 days ago) **explicitly removed** that fallback once the indexer caught up. Re-adding a guard here would revert that decision.

Deploy order for this PR, pinned to the PR description: (1) ship the schema + handler changes to the indexer and wait for Envio to expose the new fields in Hasura, (2) merge this PR so `metrics-bridge` auto-deploys against the ready schema. Not merging until step 1 is confirmed.

File: terraform/alerts/rules-fpmms.tf [hunk 6]
```diff
@@ -839,3 +839,217 @@ resource "grafana_rule_group" "fpmms_rebalancer" {
...
+
+  rule {
+    name           = "Oracle Jump Exceeds Swap Fee"
+    condition      = "threshold"
+    for            = "0m"
+    exec_err_state = "Error"
+    no_data_state  = "OK"
+
+    annotations = {
+      summary     = "Oracle jumped {{ printf \"%.2f\" $values.A.Value }} bps — above the pool's {{ printf \"%.0f\" $values.Fee.Value }} bps swap fee. LPs leaking per arb round-trip."
```
[the remark this answers is not in the document]
Comment (chapati23) [reply terraform/alerts/rules-fpmms.tf]: Fixed in b271994 — wrapped both `$values.Fee.Value` references with `{{ if $values.Fee }}...{{ else }}?{{ end }}`. Matches the existing guard pattern on Oracle Liveness (`{{ if and $values.OracleTs ... }}`) and Rebalancer Stale. If Fee is absent mid-scrape, the annotation renders `pool's ? bps swap fee` instead of template-panicking.

File: metrics-bridge/src/graphql.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (chapati23) [reply metrics-bridge/src/graphql.ts:31]: Won't fix — duplicate of the Cursor and Codex findings on this same line (resolved in their threads).

Repo discipline is indexer-first-then-bridge, not dual-read fallback: PR #212 added that fallback for `lastEffectivenessRatio`; PR #219 (2 days ago) explicitly removed it. Re-adding would revert that decision. Deploy order pinned to the PR description; not merging until the indexer schema is live on Envio.
